File: src-api/EduvibeBackend/src/main/java/com/example/EduvibeBackend/service/ClaseService.java
package com.example.EduvibeBackend.service;

import java.util.List;

import com.example.EduvibeBackend.dto.ClaseDto;
import com.example.EduvibeBackend.entities.Clase;

/**
 * Servicio para gestionar las operaciones relacionadas con las clases.
 */
public interface ClaseService {
    
    /**
     * Crea una nueva clase.
     * 
     * @param claseDto La información de la clase a crear.
     * @return La clase creada.
     */
    ClaseDto crearClase(ClaseDto claseDto);
    
    /**
     * Obtiene una clase por su identificador único.
     * 
     * @param id El identificador único de la clase.
     * @return La clase encontrada, o null si no se encontró ninguna clase con el ID dado.
     */
    ClaseDto obtenerClasePorId(Long id);
    
    /**
     * Edita una clase existente.
     * 
     * @param id El identificador único de la clase a editar.
     * @param claseDto La información actualizada de la clase.
     * @return La clase actualizada.
     */
    ClaseDto editarClase(Long id, ClaseDto claseDto);
    
    /**
     * Elimina una clase por su identificador único.
     * 
     * @param id El identificador único de la clase a eliminar.
     */
    void eliminarClase(Long id);
    
    /**
     * Obtiene todas las clases disponibles.
     * 
     * @return Una lista de todas las clases.
     */
    List<ClaseDto> obtenerTodasLasClases();
    
    /**
     * Obtiene una clase por su identificador único, sin mapear a un DTO.
     * 
     * @param id El identificador único de la clase.
     * @return La clase encontrada, o null si no se encontró ninguna clase con el ID dado.
     */
    Clase obtenerClaseSinDto(Long id);
    
    /**
     * Inscribir a un usuario en una clase.
     * 
     * @param email El correo electrónico del usuario a inscribir.
     * @param idClase El identificador único de la clase a la que se inscribirá el usuario.
     */
    void inscribirUsuarioEnClase(String email, Long idClase);
    
    /**
     * Elimina a un usuario de una clase.
     * 
     * @param email El correo electrónico del usuario a eliminar.
     * @param idClase El identificador único de la clase de la que se eliminará el usuario.
     */
    void eliminarUsuarioDeClase(String email, Long idClase);
    
    /**
     * Obtiene todas las clases a las que está inscrito un usuario.
     * 
     * @param idUsuario El identificador único del usuario.
     * @return Una lista de todas las clases a las que está inscrito el usuario.
     */
    List<ClaseDto> obtenerClasesPorUsuario(Integer idUsuario);
}

File: src-frontend/EduVibeFront/src/app/services/clase.service.ts
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { BehaviorSubject, Observable, of, throwError } from 'rxjs';
import { tap, catchError, map } from 'rxjs/operators';
import { Router } from '@angular/router';
import Swal from 'sweetalert2';
import { ClaseDto } from '../interfaces/clase';
import { UserResp } from '../interfaces/userResp';

@Injectable({
  providedIn: 'root'
})
export class ClaseService {
  private baseUrl = 'http://localhost:9090/clases'; // URL base de tu backend

  constructor(private http: HttpClient) { }

  // Crear una nueva clase
  crearClase(clase: ClaseDto): Observable<ClaseDto> {
    return this.http.post<ClaseDto>(`${this.baseUrl}/crear`, clase);
  }

  // Obtener una clase por su ID
  obtenerClasePorId(id: number): Observable<ClaseDto> {
    return this.http.get<ClaseDto>(`${this.baseUrl}/${id}`);
  }

  // Editar una clase existente
  editarClase(id: number, clase: ClaseDto): Observable<ClaseDto> {
    return this.http.put<ClaseDto>(`${this.baseUrl}/editar/${id}`, clase);
  }

  // Eliminar una clase por su ID
  eliminarClase(id: number): Observable<void> {
    return this.http.delete<void>(`${this.baseUrl}/eliminar/${id}`);
  }
  
  // Obtener todas las clases
  obtenerTodasLasClases(): Observable<ClaseDto[]> {
    return this.http.get<ClaseDto[]>(`${this.baseUrl}/todos`);
  }

  // Obtener clases por usuario
  obtenerClasesPorUsuario(idUsuario: number): Observable<ClaseDto[]> {
    return this.http.get<ClaseDto[]>(`${this.baseUrl}/usuario/${idUsuario}`);
  }
  eliminarUsuarioDeClase(email: string, idClase: number): Observable<void> {
    return this.http.delete<void>(`${this.baseUrl}/${idClase}/eliminar-usuario`, { params: { email } }).pipe(
      catchError(error => {
        console.error('Error al eliminar usuario de la clase', error);
        return throwError(error);
      })
    );
  }
  
  // Método para inscribir un usuario en una clase
  inscribirUsuario(idUsuario: number, idClase: number): Observable<void> {
    return this.http.post<void>(`${this.baseUrl}/inscribir`, null, { params: { idUsuario: idUsuario, idClase: idClase } });
  }


}
